### PR TITLE
Use AsyncCollection in shared helper

### PR DESCRIPTION
**Pull request by pakrym**
Add support for it in the async-to-sync interceptor.

### CODE REVIEW

File: sdk/core/Azure.Core/src/AsyncCollection.cs [hunk 1]
```diff
@@ -66,7 +66,16 @@ public abstract class AsyncCollection<T> : IAsyncEnumerable<Response<T>>
         /// enumerating asynchronously.
         /// </param>
         /// <returns>An async sequence of values.</returns>
-        public abstract IAsyncEnumerator<Response<T>> GetAsyncEnumerator(CancellationToken cancellationToken = default);
+        public virtual async IAsyncEnumerator<Response<T>> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            await foreach (Page<T> page in ByPage())
```
Comment (tg-msft) [inline sdk/core/Azure.Core/src/AsyncCollection.cs]: Shouldn't we thread the cancellationToken into ByPage?  I had an internal GetNextPageAsync to handle this in https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Common/src/StorageAsyncCollection.cs

**Review by tg-msft — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyClient.cs [hunk 2]
```diff
@@ -404,7 +404,7 @@ public virtual IEnumerable<Response<KeyBase>> GetKeys(CancellationToken cancella
         /// permission.
         /// </remarks>
         /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
-        public virtual IAsyncEnumerable<Response<KeyBase>> GetKeysAsync(CancellationToken cancellationToken = default)
+        public virtual AsyncCollection<KeyBase> GetKeysAsync(CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyClient.cs]: I don't think it is AsyncCollection implements IAsyncEnumerable<Response<>> or am I missing something?

[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyClient.cs]: Yes, you're right.  I used to work on a team that would have considered this a possible breaking change, but I'm of the opinion they bent over backwards a little too far in accommodating customers _(in a nutshell, if we returned a `List<T>` as an `IEnumerable` and a major customer realized it, cast it to `List<T>`, and relied on private implementation details, we would continue to support it...)_.

File: sdk/storage/Azure.Storage.Files/tests/FileClientTests.cs [hunk 3]
```diff
@@ -690,12 +690,7 @@ public async Task ListHandles()
             using (this.GetNewFile(out var file))
             {
                 // Act
-                var handles =
-                    (await file.GetHandlesAsync()
-                    .ByPage(pageSizeHint: 5)
-                    .ToListAsync())
-                    .SelectMany(p => p.Values)
-                    .ToList();
+                var handles = await file.GetHandlesAsync().ToListAsync();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files/tests/FileClientTests.cs]: Yeah...  I left some of the `ByPage` calls to keep the tests as close as possible, but this one was a bit much.

File: sdk/core/Azure.Core/tests/TestFramework/AsyncOnlyAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        /// Modifies a test by marking it as Ignored.
+        /// </summary>
+        /// <param name="test">The test to modify</param>
+        public void ApplyToTest(Test test)
+        {
+            if (test.RunState != RunState.NotRunnable)
+            {
+                // This is an unfortunate implementation but it's the only one I was able to figure out
+                string testParameters = test.FullName.Substring(test.ClassName.Length);
+                if (testParameters.StartsWith("(False"))
```
Comment (tg-msft) [inline sdk/core/Azure.Core/tests/TestFramework/AsyncOnlyAttribute.cs]: Worth filing a bug to dig into further?

**Review by tg-msft — APPROVED** [review]

File: sdk/core/Azure.Core/tests/TestFramework/AsyncOnlyAttribute.cs (hunk 4 above)
Comment (pakrym) [reply sdk/core/Azure.Core/tests/TestFramework/AsyncOnlyAttribute.cs]: Not sure there's a lot to dig into, this information doesn't seem to be on the test object, unfortunately. Let's see if it causes any issues like this.

**Pull request merged**